### DISCUSSION TITLE
Order query results by nested JSON property - ORM model

### DISCUSSION THREAD

**Discussion by nikosyal** [sqlalchemy/sqlalchemy#11150] [Usage Questions]
[SQLAlchemy version 1.4.35]
Using the ORM model, I am trying to order a table's records based on the values of the **nested** **property** contained within the custom JSON type `MyType`.

```
class MyTable:
    __tablename__ = 'my_table'
    myfield = Column('my_field', type_='MyType', nullable=False) 
```
Where the definition of `MyType` is roughly the following:

```
from sqlalchemy.dialects import postgresql, sqlite
from sqlalchemy.types import Text, TypeDecorator

class MyType(TypeDecorator):
    def load_dialect_impl(self, dialect):
        if dialect.name == 'postgresql':
            return dialect.type_descriptor(postgresql.JSONB())
        if dialect.name == 'sqlite':
            return dialect.type_descriptor(sqlite.JSON())
    # -----------------
    # rest of logic
    # -----------------
```
The end-goal is to perform something analogous to:
```
from sqlalchemy import create_engine
with Session(create_engine()) as session:
     query = session.query(schema.MyTable.MyType).order_by(
         schema.MyTable.MyType.my_field.{{my_nested_property}})
     result = query.all()
```
where `my_nested_property` is contained within the JSON MyType object. When attempted, the output is 
`AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with schema.MyTable.MyType has an attribute 'my_nested_property'`.

Unfortunately, I can not find a well-documented way to access the nested properties from within the JSON field to perform sorting.

Any suggestions?
Thank you!

**Comment by CaselIT** [comment]
Hi,

You need to use the getitem form, so `schema.MyTable.my_field[my_nested_property]`
The documentation is linked to any json type. See https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON

**Reply by MiltiadisKoutsokeras** [reply]
Unfortunately your proposal raises the exception:
```
Unhandled exception! Details: Operator 'getitem' is not supported on this expression
```

The column used by @nikosyal is implemented with this type decoration:

```Python
import json
from sqlalchemy.dialects import postgresql, sqlite
from sqlalchemy.types import Text, TypeDecorator

class AgnosticJSON(TypeDecorator):
    impl = Text
    cache_ok = True
    python_type = dict

    def load_dialect_impl(self, dialect):
        if dialect.name == 'postgresql':
            return dialect.type_descriptor(postgresql.JSONB())
        if dialect.name == 'sqlite':
            return dialect.type_descriptor(sqlite.JSON())
        return dialect.type_descriptor(self.impl)

    def process_bind_param(self, value, dialect):
        if dialect.name == 'postgresql' or value is None:
            return value
        if dialect.name == 'sqlite' or value is None:
            return value
        return json.dumps(value, sort_keys=True)

    def process_result_value(self, value, dialect):
        if dialect.name == 'postgresql' or value is None:
            return value
        if dialect.name == 'sqlite' or value is None:
            return value
        return json.loads(value)

    def process_literal_param(self, value, dialect):
        return value
```

**Reply by CaselIT** [reply]
you probably need to tell the `TypeDecorator` to behave like the json column.

Does changing `impl` to [`JSON`](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON) work? If not you will need to manually change the comparator probably using this https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.comparator_factory

**Reply by MiltiadisKoutsokeras** [reply]
The `impl=Text` is set for possible future dialects that may not support JSON columns. It is mainly used for testing compatibility. From my understanding, the only way for ORM to use our AgnosticJSON column's nested property for query (in our case ordering) is to expand it like the following:

```Python
class OurTable:
  __tablename__ = 'our_table'
  col_str = Column('col_str', type_=String, nullable=False)
  col_jsonb = Column('col_jsonb', type_=AgnosticJSON, nullable=False)
  
query = session.query(schema.OurTable).where(schema.OurTable.col_str == 'some value')
sort_column = func.jsonb_extract_path_text(schema.OurTable.col_jsonb, 'JSON_Key')
query = query.order_by(sort_column)
query.all()
```

This unfortunately only works on Postgres where the function for extraction exists and breaks in all other cases.

**Reply by CaselIT** [reply]
> The impl=Text is set for possible future dialects that may not support JSON columns

You could use `return dialect.type_descriptor(Text)`

> From my understanding, the only way for ORM to use our AgnosticJSON column's nested property for query (in our case ordering) is to expand it like the following:

No, you can tell the type to use the json comparator like I mentioned above. The provided link has an example that similar to what you want to do.

**Reply by MiltiadisKoutsokeras** [reply]
Thanks for the information. Is there an example that we could look for the comparator factory?

**Reply by CaselIT** [reply]
see this example that uses the integer comparator https://docs.sqlalchemy.org/en/20/core/custom_types.html#types-operators

**Comment by zzzeek** [comment]
hi -

We have [with_variant](https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant) for this use case:

```py
class MyTable:
    __tablename__ = 'my_table'
    myfield = Column('my_field', JSON().with_variant(postgresql.JSONB(), "postgresql"), nullable=False) 
```

the JSON operators should work in 2.0 for with_variant(), not sure about 1.4.

**Reply by MiltiadisKoutsokeras** [reply]
I implemented this type decoration class logic back when SQLAlchemy 1.4.35 was released.

**Reply by zzzeek** [reply]
getting the JSON operators to work with TypeDecorator might be tricky, though.

**Reply by CaselIT** [reply]
isn't it just a question of setting the correct comparator?
